[mlir] Lower math dialect later in gpu-lower-to-nvvm-pipeline

grypp opened the pull request:
This PR moves lowering of math dialect later in the pipeline. Because math dialect is lowered correctly by `createConvertGpuOpsToNVVMOps` for GPU target, and it needs to run it first.

apaszke commented on stmatrix/everything-casted.mlir:
```diff
@@ -0,0 +1,1267 @@
+// -----// IR Dump After ConvertNVGPUToNVVMPass (convert-nvgpu-to-nvvm) //----- //
```
I assume this is an accidental inclusion to this PR?

apaszke commented on mlir/test/Dialect/GPU/test-nvvm-pipeline.mlir:
```diff
@@ -0,0 +1,18 @@
+    gpu.launch 
+        blocks(%0, %1, %2) in (%3 = %c1, %4 = %c1, %5 = %c1) 
+        threads(%6, %7, %8) in (%9 = %c2, %10 = %c1, %11 = %c1) { 
+         %s1 = math.exp %arg0 : f32
```
Add a check that there is a call to the libdevice function?

apaszke commented on mlir/test/Dialect/GPU/test-nvvm-pipeline.mlir:
```diff
@@ -0,0 +1,18 @@
+// CHECK-LABEL: llvm.func @test_math(%arg0: f32) {
+func.func @test_math(%arg0 : f32) {
+    %c2 = arith.constant 2 : index
+    %c1 = arith.constant 1 : index
```
Perhaps add an exp here as well and check that it lowers to llvm intrinsics?

grypp committed:
[mlir] Lower math dialect later in gpu-lower-to-nvvm-pipeline

This PR moves lowering of math dialect later in the pipeline. Because math dialect is lowered correctly by `createConvertGpuOpsToNVVMOps` for GPU target, and it needs to run it first.

grypp committed:
check the right thing

grypp commented:
Buildbot did not like the way I check PTX. I'll fix this quickly.
